src/sidebar/components/test/tag-editor-test.js

```diff
@@ -1,63 +1,257 @@
+      assert.equal(tag.prop('aria-label'), `Tag: ${fakeTags[i]}`);
     });
   });
 
-  it('converts tags to the form expected by ng-tags-input', function() {
-    const element = util.createDirective(document, 'tag-editor', {
-      tags: ['foo', 'bar'],
+  it('creates the component with a unique id', () => {
+    const wrapper = createComponent();
+    assert.equal(
+      wrapper.find('input').prop('list'),
+      'tag-editor-datalist-annotation_id'
+    );
+  });
+
+  it('calls fakeTagsService.filter with the value from the input field', () => {
+    const wrapper = createComponent();
+    wrapper.find('input').instance().value = 'tag3';
+    // simulate `keyup` to populate suggestions list
+    wrapper.find('input').simulate('keyup', { key: 'any' });
+    assert.isTrue(fakeTagsService.filter.calledWith('tag3'));
+  });
+
+  it('generates a datalist set equal to the array value returned from fakeTagsService.filter ', () => {
+    const wrapper = createComponent();
+    // simulate `keyup` to populate suggestions list
+    wrapper.find('input').simulate('keyup', { key: 'any' });
+
+    assert.equal(
+      wrapper
+        .find('datalist option')
+        .at(0)
+        .prop('value'),
+      'tag3'
+    );
+    assert.equal(
+      wrapper
+        .find('datalist option')
+        .at(1)
+        .prop('value'),
+      'tag4'
+    );
+  });
+
+  it('clears the suggestions when adding a new tag', () => {
+    const wrapper = createComponent();
+    wrapper.find('input').instance().value = 'tag3';
+    // simulate `keyup` to populate suggestions list
+    wrapper.find('input').simulate('keyup', { key: 'any' });
+    assert.equal(wrapper.find('datalist option').length, 2);
+    wrapper.find('input').simulate('keyup'); // no key supplied
+    assert.equal(wrapper.find('datalist option').length, 0);
+
+    assert.isTrue(
+      fakeTagsService.store.calledWith([
+        { text: 'tag1' },
+        { text: 'tag2' },
+        { text: 'tag3' },
+      ])
+    );
+  });
+
+  it('does not render duplicate suggestions', () => {
+    // `tag3` supplied in the `tagList` will be a duplicate value relative
+    // with the fakeTagsService.filter result above.
+    const wrapper = createComponent({
+      editMode: true,
+      tagList: ['tag1', 'tag2', 'tag3'],
     });
-    assert.deepEqual(element.ctrl.tagList, [{ text: 'foo' }, { text: 'bar' }]);
+    // simulate `keyup` to populate suggestions list
+    wrapper.find('input').simulate('keyup', { key: 'any' });
+    assert.equal(wrapper.find('datalist option').length, 1);
+    assert.equal(
+      wrapper
+        .find('datalist option')
+        .at(0)
+        .prop('value'),
+      'tag4'
+    );
   });
 
-  describe('when tags are changed', function() {
-    let element;
-    let onEditTags;
+  context('when adding tags', () => {
+    /**
+     * Helper function to assert that a tag was correctly added
+     */
+    const addTagsSuccess = (wrapper, tagList) => {
```

Comment:
I wonder if the name of this function should reflect that there are embedded assertions. I don't know that we have a convention for this, offhand.

Reply:
How about we prefix it with assert?